File: src/mffs/MFFSHelper.java
package mffs;

import icbm.api.IBlockFrequency;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Set;

import mffs.api.IProjector;
import mffs.api.fortron.IFortronFrequency;
import mffs.api.modules.IModuleAcceptor;
import mffs.api.security.IInterdictionMatrix;
import mffs.api.security.Permission;
import mffs.fortron.FrequencyGrid;
import mffs.item.module.projector.ItemModeCustom;
import net.minecraft.block.Block;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.inventory.IInventory;
import net.minecraft.item.ItemBlock;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.world.World;
import net.minecraftforge.common.ForgeDirection;
import net.minecraftforge.event.entity.player.PlayerInteractEvent.Action;
import universalelectricity.core.vector.Vector3;
import calclavia.lib.CalculationHelper;

/**
 * A class containing some general helpful functions.
 * 
 * @author Calclavia
 * 
 */
public class MFFSHelper
{
	public static void transferFortron(IFortronFrequency transferer, Set<IFortronFrequency> frequencyTiles, TransferMode transferMode, int limit)
	{
		if (transferer != null && frequencyTiles.size() > 1)
		{
			/**
			 * Check spread mode. Equal, Give All, Take All
			 */
			int totalFortron = 0;
			int totalCapacity = 0;

			for (IFortronFrequency machine : frequencyTiles)
			{
				if (machine != null)
				{
					totalFortron += machine.getFortronEnergy();
					totalCapacity += machine.getFortronCapacity();
				}
			}

			if (totalFortron > 0 && totalCapacity > 0)
			{
				/**
				 * Test each mode and based on the mode, spread Fortron energy.
				 */
				switch (transferMode)
				{
					case EQUALIZE:
					{
						for (IFortronFrequency machine : frequencyTiles)
						{
							if (machine != null)
							{
								double capacityPercentage = (double) machine.getFortronCapacity() / (double) totalCapacity;
								int amountToSet = (int) (totalFortron * capacityPercentage);
								doTransferFortron(transferer, machine, amountToSet - machine.getFortronEnergy(), limit);
							}
						}

						break;
					}
					case DISTRIBUTE:
					{
						final int amountToSet = totalFortron / frequencyTiles.size();

						for (IFortronFrequency machine : frequencyTiles)
						{
							if (machine != null)
							{
								doTransferFortron(transferer, machine, amountToSet - machine.getFortronEnergy(), limit);
							}
						}

						break;
					}
					case DRAIN:
					{
						frequencyTiles.remove(transferer);

						for (IFortronFrequency machine : frequencyTiles)
						{
							if (machine != null)
							{
								double capacityPercentage = (double) machine.getFortronCapacity() / (double) totalCapacity;
								int amountToSet = (int) (totalFortron * capacityPercentage);

								if (amountToSet - machine.getFortronEnergy() > 0)
								{
									doTransferFortron(transferer, machine, amountToSet - machine.getFortronEnergy(), limit);
								}
							}
						}

						break;
					}
					case FILL:
					{
						if (transferer.getFortronEnergy() < transferer.getFortronCapacity())
						{
							frequencyTiles.remove(transferer);

							// The amount of energy required to be full.
							int requiredFortron = transferer.getFortronCapacity() - transferer.getFortronEnergy();

							for (IFortronFrequency machine : frequencyTiles)
							{
								if (machine != null)
								{
									int amountToConsume = Math.min(requiredFortron, machine.getFortronEnergy());
									int amountToSet = -machine.getFortronEnergy() - amountToConsume;

									if (amountToConsume > 0)
									{
										doTransferFortron(transferer, machine, amountToSet - machine.getFortronEnergy(), limit);
									}
								}
							}
						}

						break;
					}
				}
			}
		}
	}

	/**
	 * Tries to transfer Fortron to a specific machine from this capacitor. Renders an animation on
	 * the client side.
	 * 
	 * @param receiver: The machine to be transfered to.
	 * @param joules: The amount of energy to be transfered.
	 */
	public static void doTransferFortron(IFortronFrequency transferer, IFortronFrequency receiver, int joules, int limit)
	{
		if (transferer != null && receiver != null)
		{
			TileEntity tileEntity = ((TileEntity) transferer);
			World world = tileEntity.worldObj;

			boolean isCamo = false;

			if (transferer instanceof IModuleAcceptor)
			{
				isCamo = ((IModuleAcceptor) transferer).getModuleCount(ModularForceFieldSystem.itemModuleCamouflage) > 0;
			}

			if (joules > 0)
			{
				// Transfer energy to receiver.
				joules = Math.min(joules, limit);
				int toBeInjected = receiver.provideFortron(transferer.requestFortron(joules, false), false);
				toBeInjected = transferer.requestFortron(receiver.provideFortron(toBeInjected, true), true);

				// Draw Beam Effect
				if (world.isRemote && toBeInjected > 0 && !isCamo)
				{
					ModularForceFieldSystem.proxy.renderBeam(world, Vector3.add(new Vector3(tileEntity), 0.5), Vector3.add(new Vector3((TileEntity) receiver), 0.5), 0.6f, 0.6f, 1, 20);
				}
			}
			else
			{
				// Take energy from receiver.
				joules = Math.min(Math.abs(joules), limit);
				int toBeEjected = transferer.provideFortron(receiver.requestFortron(joules, false), false);
				toBeEjected = receiver.requestFortron(transferer.provideFortron(toBeEjected, true), true);

				// Draw Beam Effect
				if (world.isRemote && toBeEjected > 0 && !isCamo)
				{
					ModularForceFieldSystem.proxy.renderBeam(world, Vector3.add(new Vector3((TileEntity) receiver), 0.5), Vector3.add(new Vector3(tileEntity), 0.5), 0.6f, 0.6f, 1, 20);
				}

			}
		}

	}

	/**
	 * Gets the nearest active Interdiction Matrix.
	 */
	public static IInterdictionMatrix getNearestInterdictionMatrix(World world, Vector3 position)
	{
		for (IBlockFrequency frequencyTile : FrequencyGrid.instance().get())
		{
			if (((TileEntity) frequencyTile).worldObj == world && frequencyTile instanceof IInterdictionMatrix)
			{
				IInterdictionMatrix interdictionMatrix = (IInterdictionMatrix) frequencyTile;

				if (interdictionMatrix.isActive())
				{
					if (position.distanceTo(new Vector3((TileEntity) interdictionMatrix)) <= interdictionMatrix.getActionRange())
					{
						return interdictionMatrix;
					}
				}
			}
		}

		return null;
	}

	/**
	 * Returns true of the interdictionMatrix has a specific set of permissions.
	 * 
	 * @param interdictionMatrix
	 * @param username
	 * @param permissions
	 * @return
	 */
	public static boolean isPermittedByInterdictionMatrix(IInterdictionMatrix interdictionMatrix, String username, Permission... permissions)
	{
		if (interdictionMatrix != null)
		{
			if (interdictionMatrix.isActive())
			{
				if (interdictionMatrix.getBiometricIdentifier() != null)
				{
					for (Permission permission : permissions)
					{
						if (!interdictionMatrix.getBiometricIdentifier().isAccessGranted(username, permission))
						{
							if (interdictionMatrix.getModuleCount(ModularForceFieldSystem.itemModuleInvert) > 0)
							{
								return true;
							}
							else
							{
								return false;
							}
						}
					}
				}
			}
		}

		if (interdictionMatrix.getModuleCount(ModularForceFieldSystem.itemModuleInvert) > 0)
		{
			return false;
		}
		else
		{
			return true;
		}
	}

	public static List<String> splitStringPerWord(String string, int wordsPerLine)
	{
		String[] words = string.split(" ");
		List<String> lines = new ArrayList<String>();

		for (int lineCount = 0; lineCount < Math.ceil((float) words.length / (float) wordsPerLine); lineCount++)
		{
			String stringInLine = "";

			for (int i = lineCount * wordsPerLine; i < Math.min(wordsPerLine + lineCount * wordsPerLine, words.length); i++)
			{
				stringInLine += words[i] + " ";
			}

			lines.add(stringInLine.trim());
		}

		return lines;
	}

	/**
	 * Gets the first itemStack that is an ItemBlock in this TileEntity or in nearby chests.
	 * 
	 * @param itemStack
	 * @return
	 */
	public static ItemStack getFirstItemBlock(TileEntity tileEntity, ItemStack itemStack)
	{
		return getFirstItemBlock(tileEntity, itemStack, true);
	}

	public static ItemStack getFirstItemBlock(TileEntity tileEntity, ItemStack itemStack, boolean recur)
	{
		if (tileEntity instanceof IProjector)
		{
			for (int i : ((IProjector) tileEntity).getModuleSlots())
			{
				ItemStack checkStack = getFirstItemBlock(i, ((IProjector) tileEntity), itemStack);

				if (checkStack != null)
				{
					return checkStack;
				}
			}
		}
		else if (tileEntity instanceof IInventory)
		{
			IInventory inventory = (IInventory) tileEntity;

			for (int i = 0; i < inventory.getSizeInventory(); i++)
			{
				ItemStack checkStack = getFirstItemBlock(i, inventory, itemStack);

				if (checkStack != null)
				{
					return checkStack;
				}
			}
		}

		if (recur)
		{
			for (int i = 0; i < 6; i++)
			{
				ForgeDirection direction = ForgeDirection.getOrientation(i);
				Vector3 vector = new Vector3(tileEntity);
				vector.modifyPositionFromSide(direction);
				TileEntity checkTile = vector.getTileEntity(tileEntity.worldObj);

				if (checkTile != null)
				{
					ItemStack checkStack = getFirstItemBlock(checkTile, itemStack, false);

					if (checkStack != null)
					{
						return checkStack;
					}
				}
			}
		}

		return null;
	}

	public static ItemStack getFirstItemBlock(int i, IInventory inventory, ItemStack itemStack)
	{
		ItemStack checkStack = inventory.getStackInSlot(i);

		if (checkStack != null && checkStack.getItem() instanceof ItemBlock)
		{
			if (itemStack == null || checkStack.isItemEqual(itemStack))
			{
				return checkStack;
			}
		}
		return null;
	}

	public static Block getFilterBlock(ItemStack itemStack)
	{
		if (itemStack != null)
		{
			if (itemStack.getItem() instanceof ItemBlock)
			{
				if (((ItemBlock) itemStack.getItem()).getBlockID() < Block.blocksList.length)
				{
					Block block = Block.blocksList[((ItemBlock) itemStack.getItem()).getBlockID()];

					if (block.renderAsNormalBlock())
					{
						return block;
					}
				}
			}
		}

		return null;

	}

	public static ItemStack getCamoBlock(IProjector projector, Vector3 position)
	{
		if (projector != null)
		{
			if (!((TileEntity) projector).worldObj.isRemote)
			{
				if (projector != null)
				{
					if (projector.getModuleCount(ModularForceFieldSystem.itemModuleCamouflage) > 0)
					{
						if (projector.getMode() instanceof ItemModeCustom)
						{
							HashMap<Vector3, int[]> fieldMap = ((ItemModeCustom) projector.getMode()).getFieldBlockMap(projector, projector.getModeStack());

							if (fieldMap != null)
							{
								Vector3 fieldCenter = new Vector3((TileEntity) projector).add(projector.getTranslation());
								Vector3 relativePosition = position.clone().subtract(fieldCenter);
								CalculationHelper.rotateByAngle(relativePosition, -projector.getRotationYaw(), -projector.getRotationPitch());
								int[] blockInfo = fieldMap.get(relativePosition.round());

								if (blockInfo != null && blockInfo[0] > 0)
								{
									return new ItemStack(Block.blocksList[blockInfo[0]], 1, blockInfo[1]);
								}
							}
						}

						for (int i : projector.getModuleSlots())
						{
							ItemStack checkStack = projector.getStackInSlot(i);
							Block block = getFilterBlock(checkStack);

							if (block != null)
							{
								return checkStack;
							}
						}
					}
				}
			}
		}

		return null;
	}

	/**
	 * Gets a compound from an itemStack.
	 * 
	 * @param itemStack
	 * @return
	 */
	public static NBTTagCompound getNBTTagCompound(ItemStack itemStack)
	{
		if (itemStack != null)
		{
			if (itemStack.getTagCompound() == null)
			{
				itemStack.setTagCompound(new NBTTagCompound());
			}

			return itemStack.getTagCompound();
		}

		return null;
	}

	public static boolean hasPermission(World world, Vector3 position, Permission permission, EntityPlayer player)
	{
		IInterdictionMatrix interdictionMatrix = getNearestInterdictionMatrix(world, position);

		if (interdictionMatrix != null)
		{
			return isPermittedByInterdictionMatrix(interdictionMatrix, player.username, permission);
		}

		return true;
	}

	public static boolean hasPermission(World world, Vector3 position, Action action, EntityPlayer player)
	{
		IInterdictionMatrix interdictionMatrix = getNearestInterdictionMatrix(world, position);

		if (interdictionMatrix != null)
		{
			return MFFSHelper.hasPermission(world, position, interdictionMatrix, action, player);
		}

		return true;
	}

	public static boolean hasPermission(World world, Vector3 position, IInterdictionMatrix interdictionMatrix, Action action, EntityPlayer player)
	{
		boolean hasPermission = true;

		if (action == Action.RIGHT_CLICK_BLOCK && position.getTileEntity(world) != null)
		{
			if (interdictionMatrix.getModuleCount(ModularForceFieldSystem.itemModuleBlockAccess) > 0)
			{
				hasPermission = false;

				if (isPermittedByInterdictionMatrix(interdictionMatrix, player.username, Permission.BLOCK_ACCESS))
				{
					hasPermission = true;
				}
			}
		}

		if (hasPermission)
		{
			if (interdictionMatrix.getModuleCount(ModularForceFieldSystem.itemModuleBlockAlter) > 0 && (player.getCurrentEquippedItem() != null || action == Action.LEFT_CLICK_BLOCK))
			{
				hasPermission = false;

				if (isPermittedByInterdictionMatrix(interdictionMatrix, player.username, Permission.BLOCK_ALTER))
				{
					hasPermission = true;
				}
			}
		}

		return hasPermission;
	}

}

File: src/main/scala/mffs/MFFSHelper.scala
package mffs

import com.mojang.authlib.GameProfile
import mffs.field.TileElectromagnetProjector
import mffs.field.mode.ItemModeCustom
import mffs.fortron.TransferMode
import mffs.fortron.TransferMode.TransferMode
import net.minecraft.block.Block
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.inventory.IInventory
import net.minecraft.item.{ItemBlock, ItemStack}
import net.minecraft.tileentity.TileEntity
import net.minecraft.world.World
import net.minecraftforge.common.util.ForgeDirection
import net.minecraftforge.event.entity.player.PlayerInteractEvent
import resonant.api.mffs.IProjector
import resonant.api.mffs.fortron.{FrequencyGridRegistry, IFortronFrequency}
import resonant.api.mffs.modules.IModuleAcceptor
import resonant.api.mffs.security.IInterdictionMatrix
import resonant.engine.grid.frequency.FrequencyGrid
import resonant.lib.access.Permission
import universalelectricity.core.transform.rotation.Rotation
import universalelectricity.core.transform.vector.Vector3

import scala.collection.JavaConversions._
import scala.collection.mutable

/**
 * A class containing some general helpful functions.
 *
 * @author Calclavia
 */
object MFFSHelper
{
  def transferFortron(transferer: IFortronFrequency, frequencyTiles: Set[IFortronFrequency], transferMode: TransferMode, limit: Int)
  {
    if (transferer != null && frequencyTiles.size > 1 && Settings.allowFortronTeleport)
    {
      var totalFortron: Int = 0
      var totalCapacity: Int = 0

      for (machine <- frequencyTiles)
      {
        if (machine != null)
        {
          totalFortron += machine.getFortronEnergy
          totalCapacity += machine.getFortronCapacity
        }
      }
      if (totalFortron > 0 && totalCapacity > 0)
      {
        transferMode match
        {
          case TransferMode.EQUALIZE =>
          {
            for (machine <- frequencyTiles)
            {
              if (machine != null)
              {
                val capacityPercentage: Double = machine.getFortronCapacity.asInstanceOf[Double] / totalCapacity.asInstanceOf[Double]
                val amountToSet: Int = (totalFortron * capacityPercentage).asInstanceOf[Int]
                doTransferFortron(transferer, machine, amountToSet - machine.getFortronEnergy, limit)
              }
            }
          }
          case TransferMode.DISTRIBUTE =>
          {
            val amountToSet: Int = totalFortron / frequencyTiles.size
            for (machine <- frequencyTiles)
            {
              if (machine != null)
              {
                doTransferFortron(transferer, machine, amountToSet - machine.getFortronEnergy, limit)
              }
            }
          }
          case TransferMode.DRAIN =>
          {
            frequencyTiles.remove(transferer)

            for (machine <- frequencyTiles)
            {
              if (machine != null)
              {
                val capacityPercentage: Double = machine.getFortronCapacity.asInstanceOf[Double] / totalCapacity.asInstanceOf[Double]
                val amountToSet: Int = (totalFortron * capacityPercentage).asInstanceOf[Int]

                if (amountToSet - machine.getFortronEnergy > 0)
                {
                  doTransferFortron(transferer, machine, amountToSet - machine.getFortronEnergy, limit)
                }
              }
            }
          }
          case TransferMode.FILL =>
          {
            if (transferer.getFortronEnergy < transferer.getFortronCapacity)
            {
              frequencyTiles.remove(transferer)
              val requiredFortron: Int = transferer.getFortronCapacity - transferer.getFortronEnergy

              for (machine <- frequencyTiles)
              {
                if (machine != null)
                {
                  val amountToConsume: Int = Math.min(requiredFortron, machine.getFortronEnergy)
                  val amountToSet: Int = -machine.getFortronEnergy - amountToConsume
                  if (amountToConsume > 0)
                  {
                    doTransferFortron(transferer, machine, amountToSet - machine.getFortronEnergy, limit)
                  }
                }
              }
            }
          }
        }
      }
    }
  }

  /**
   * Tries to transfer Fortron to a specific machine from this capacitor. Renders an animation on
   * the client side.
   *
   * @param receiver : The machine to be transfered to.
   * @param joules   : The amount of energy to be transfered.
   */
  def doTransferFortron(transferer: IFortronFrequency, receiver: IFortronFrequency, joules: Int, limit: Int)
  {
    if (transferer != null && receiver != null)
    {
      val tileEntity = transferer.asInstanceOf[TileEntity]
      val world: World = tileEntity.getWorldObj()
      var isCamo = false

      if (transferer.isInstanceOf[IModuleAcceptor])
      {
        isCamo = (transferer.asInstanceOf[IModuleAcceptor]).getModuleCount(ModularForceFieldSystem.Items.moduleCamouflage) > 0
      }

      if (joules > 0)
      {
        val transferEnergy = Math.min(joules, limit)
        var toBeInjected: Int = receiver.provideFortron(transferer.requestFortron(transferEnergy, false), false)
        toBeInjected = transferer.requestFortron(receiver.provideFortron(toBeInjected, true), true)
        if (world.isRemote && toBeInjected > 0 && !isCamo)
        {
          ModularForceFieldSystem.proxy.renderBeam(world, new Vector3(tileEntity) + 0.5, new Vector3(receiver.asInstanceOf[TileEntity]) + 0.5, 0.6f, 0.6f, 1, 20)
        }
      }
      else
      {
        val transferEnergy = Math.min(Math.abs(joules), limit)
        var toBeEjected: Int = transferer.provideFortron(receiver.requestFortron(transferEnergy, false), false)
        toBeEjected = receiver.requestFortron(transferer.provideFortron(toBeEjected, true), true)
        if (world.isRemote && toBeEjected > 0 && !isCamo)
        {
          ModularForceFieldSystem.proxy.renderBeam(world, new Vector3(receiver.asInstanceOf[TileEntity]) + 0.5, new Vector3(tileEntity) + 0.5, 0.6f, 0.6f, 1, 20)
        }
      }
    }
  }



  /**
   * Gets the first itemStack that is an ItemBlock in this TileEntity or in nearby chests.
   */
  def getFirstItemBlock(tileEntity: TileEntity, itemStack: ItemStack): ItemStack =
  {
    return getFirstItemBlock(tileEntity, itemStack, true)
  }

  def getFirstItemBlock(tileEntity: TileEntity, itemStack: ItemStack, recur: Boolean): ItemStack =
  {
    if (tileEntity.isInstanceOf[IProjector])
    {
      val projector = tileEntity.asInstanceOf[IProjector]

      projector.getModuleSlots().find(getFirstItemBlock(_, projector, itemStack) != null) match
      {
        case Some(entry) => return getFirstItemBlock(entry, projector, itemStack)
        case _ =>
      }
    }
    else if (tileEntity.isInstanceOf[IInventory])
    {
      val inventory = tileEntity.asInstanceOf[IInventory]

      (0 until inventory.getSizeInventory()).view map (getFirstItemBlock(_, inventory, itemStack)) headOption match
      {
        case Some(entry) => return entry
        case _ =>
      }
    }

    if (recur)
    {
      ForgeDirection.VALID_DIRECTIONS.foreach(
        direction =>
        {
          val vector = new Vector3(tileEntity) + direction
          val checkTile = vector.getTileEntity(tileEntity.getWorldObj())

          if (checkTile != null)
          {
            val checkStack: ItemStack = getFirstItemBlock(checkTile, itemStack, false)

            if (checkStack != null)
            {
              return checkStack
            }
          }
        })
    }
    return null
  }

  def getFirstItemBlock(i: Int, inventory: IInventory, itemStack: ItemStack): ItemStack =
  {
    val checkStack: ItemStack = inventory.getStackInSlot(i)
    if (checkStack != null && checkStack.getItem.isInstanceOf[ItemBlock])
    {
      if (itemStack == null || checkStack.isItemEqual(itemStack))
      {
        return checkStack
      }
    }
    return null
  }

  def getFilterBlock(itemStack: ItemStack): Block =
  {
    if (itemStack != null)
    {
      if (itemStack.getItem.isInstanceOf[ItemBlock])
      {
        return itemStack.getItem().asInstanceOf[ItemBlock].field_150939_a
      }
    }
    return null
  }

  def getCamoBlock(projector: IProjector, position: Vector3): ItemStack =
  {
    val tile = projector.asInstanceOf[TileEntity]

    if (projector != null)
    {
      if (!tile.getWorldObj().isRemote)
      {
        if (projector != null)
        {
          if (projector.getModuleCount(ModularForceFieldSystem.Items.moduleCamouflage) > 0)
          {
            if (projector.getMode.isInstanceOf[ItemModeCustom])
            {
              val fieldMap = (projector.getMode.asInstanceOf[ItemModeCustom]).getFieldBlockMap(projector, projector.getModeStack)

              if (fieldMap != null)
              {
                val fieldCenter: Vector3 = new Vector3(projector.asInstanceOf[TileEntity]) + projector.getTranslation()
                var relativePosition: Vector3 = position.clone.subtract(fieldCenter)
                relativePosition = relativePosition.apply(new Rotation(-projector.getRotationYaw, -projector.getRotationPitch, 0))

                val blockInfo = fieldMap(relativePosition.round)

                if (blockInfo != null && blockInfo._1.isAir(tile.getWorldObj(), position.xi, position.yi, position.zi))
                {
                  return new ItemStack(blockInfo._1, 1, blockInfo._2)
                }
              }
            }

            for (i <- projector.getModuleSlots)
            {
              val checkStack: ItemStack = projector.getStackInSlot(i)
              val block: Block = getFilterBlock(checkStack)
              if (block != null)
              {
                return checkStack
              }
            }
          }
        }
      }
    }
    return null
  }

  /**
   * Gets the set of projectors that have an effect in this position.
   */
  def getRelevantProjectors(world: World, position: Vector3): mutable.Set[TileElectromagnetProjector] =
  {
    return FrequencyGridRegistry.instance.asInstanceOf[FrequencyGrid].getNodes(classOf[TileElectromagnetProjector]) filter (_.isInField(position))
  }

  def hasPermission(world: World, position: Vector3, permission: Permission, player: EntityPlayer): Boolean =
  {
    return hasPermission(world, position, permission, player.getGameProfile())
  }

  def hasPermission(world: World, position: Vector3, permission: Permission, profile: GameProfile): Boolean =
  {
    return getRelevantProjectors(world, position).forall(_.isAccessGranted(profile, permission))
  }

  def hasPermission(world: World, position: Vector3, action: PlayerInteractEvent.Action, player: EntityPlayer): Boolean =
  {
    return hasPermission(world, position, action, player.getGameProfile)
  }

  def hasPermission(world: World, position: Vector3, action: PlayerInteractEvent.Action, profile: GameProfile): Boolean =
  {
    return getRelevantProjectors(world, position).forall(_.isAccessGranted(world, position, profile, action))
  }
}